### ISSUE TITLE
Config examples for  bells-1.0

### ISSUE THREAD

**Issue by dmuelle** [OpenLiberty/docs#2417]
Need to work with a developer to identify 2-3 common configuration examples for [Basic Extensions using Liberty Libraries 1.0](https://www.openliberty.io/docs/20.0.0.8/reference/feature/bells-1.0.html).

- possible SME's : @una-tapa , @tjwatson 
- also investigate bells-2.0 whic h is in the code but not documented
- see user comment from #2284 (Though that particular task will be developed in another issue, might inform choice of config examples)

**Comment by tjwatson** [comment]
I think a good example to document would be a user provided implementation of SerlvetContextInitializer as discussed in https://stackoverflow.com/questions/35379315/create-a-servlet-filter-websphere-liberty-profile

Another possibility would be a customer UserRegistry, but I am not sure that is supported by liberty or not.

**Comment by dmuelle** [comment]
Thanks @tjwatson - we are actually working on developing a task topic for creating a custom user registry with bells in #2418 but any config-specifc examples for thatcher’s  task will go on the feature page we're developing here.

**Comment by ggam** [comment]
> Another possibility would be a customer UserRegistry, but I am not sure that is supported by liberty or not.

That's probably the most important part, knowing when you need a feature and when a BELL is enough. All the ibm.com documentation use features, which is overkill for me. It wasn't until these days that I found that most of those examples can be implemented just with a BELL, and that Liberty has a lot of SPI hooks available (like this one for a global JAX-WS handler https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_web_services_global_handlers.html).

This BELL thing makes extending the application server way easier that it is on WildFly, in my experience. I think it's worth documenting and promoting it.

**Comment by Rwalls1** [comment]
Link to draft: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/reference/feature/bells-1.0.html

**Comment by Rwalls1** [comment]
@tjwatson Please review: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/reference/feature/bells-1.0.html. Is there also a  basic configuration for the BELL feature?

**Comment by ggam** [comment]
@Rwalls1 while the example is correct, IMO as a user, it misses the point of explaining what exactly BELL does: register Java `ServiceLoader` classes to OSGi so that the application server can use them. One usecase is defining a `ServletContainerInitializer`, but this can be used to hook into a lot of SPI (see #2422). BELLs are an alternative to OSGi bundles for simple enough features (it will become clearer when #2418 is also done).

**Comment by tjwatson** [comment]
I agree with @ggam that the example could use a bit of an intro describing what the bells feature actually does with respect to the Java `ServiceLoader` meta files under `/META-INF/services`

I suggest something like the following:

Open Liberty features are composed of a set of OSGi bundles.  The OSGi bundles associated with an Open Liberty feature are enabled at runtime when a feature is configured with `featureManager`.  Many of the Open Liberty features can then be extended by third parties that provide OSGi service implementations for the available SPI or API provided by the Open Liberty features.  Typically this would involve implementing one or more OSGi bundles and then configuring them with the Open Liberty runtime with a user feature.  This is a powerful way to extend the Open Liberty runtime, but often it can be tedious to develop for simple scenarios.

The Bells feature provides a more simple way of extending the Open Liberty runtime by allowing a library to define an implementation of some Open Liberty API or SPI by using Java `ServiceLoader` configuration files included in the `/META-INF/services/` directory of the library JAR.  Using this approach a library declares a `/META-INF/services/` resource with the name of the fully qualified API or SPI interface.  The library also provides a class that implements this API or SPI interface.  The `/META-INF/services/` resource contains a single line that defines the class that provides the implementation of the API or SPI.  One example of an API that can be used to extend the Open Liberty runtime is the `ServletContainerInitializer`.  The bells feature can be used to process the `/META-INF/services/` resources and do the work to provide the implementation provided by the library to the Open Liberty runtime.

**Comment by Rwalls1** [comment]
@tjwatson Please review, I've added some of the intro details that you suggested regarding the feature, I didn't include the OSGI information as I think those details are more suitable for a conceptual document: 
https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/feature/bells-1.0.html

**Comment by tjwatson** [comment]
The sentence needs some work:

`The BELLs feature extends the Open Liberty runtime by allowing libraries to define the implementation of an Open Liberty SPI or API with Java ServiceLoader configuration files.`

The bells feature itself does not extend Open Liberty, rather it enables libraries to extend Open Liberty.  I would reword to this:

`The BELLs feature enables libraries to extend the Open Liberty runtime by allowing libraries to define the implementation of an Open Liberty SPI or API with Java ServiceLoader configuration files.`

**Comment by Rwalls1** [comment]
@tjwatson Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/feature/bells-1.0.html

**Comment by tjwatson** [comment]
LGTM

**Comment by Rwalls1** [comment]
@una-tapa Please review: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/feature/bells-1.0.html

**Comment by una-tapa** [comment]
Looks good to me, thanks!

**Comment by lauracowen** [comment]
Thanks. Some fairly minor comments/questions:

- Should "BELLS" be "BELLs"? The "S" doesn't seem to stand for anything? The first sentence abbreviates to "BELL" in parentheses but then "BELLS" is used subsequently - needs to be consistent.

- "Configuring the Servlet Filter" example
  - Remove "In Open Libery" - this is all about Open Liberty and can be taken as read.
  - And use "for Open Liberty" rather than "for the Open Liberty server" (we assume a single server per application so no need to separate the concept of server from OL itself usually).
  - Is it possible (in or just after the first sentence) to say why this is useful? Could give the example from the Stack Overflow question (eg to see what messages the application is sending and receiving) - but check with tech reviewer for accuracy of example and whether it's the common reason people might need this config example - and are there other examples of when/why you'd use it? Also, I think you need to mention that the point of using the BELLs feature to do this (rather than packaging it in the application as Alasdair discussed in the SO answer) is so that you don't need to modify the application (see the original post on SO). Otherwise, it's not clear why you'd use BELLs to do it.
  - Does there need to be a sentence that briefly describes what the MyFilter.class does? (not necessarily providing the actual code; just a description of what it is). I agree we don't want loads of detail in this topic (as the focus is on the configuration) but the user needs to be able to use the example. I'm not sure we'd have a separate conceptual topic on this specifically so I just want to check that this example works in its own right.

- "Adding the shared library for a custom user registry" example
  - Remove "In Open Liberty" as it's not necessary.
  - Can you check the XML tag format we should use - I think we usually close the tags like `<bell libraryRef="bellsCurLib" />` rather than like `<bell libraryRef="bellsCurLib"></bell>`. But check.

**Comment by Rwalls1** [comment]
New link to topic: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/feature/bells-1.0.html

**Comment by tjwatson** [comment]
The `Configuring the Servlet Filter` section needs some work.

The title should not use the term "the" Servlet.  That makes it sound like "the one and only".  It should be something like `Configuring a Servlet Filter for All Applications`

It could use an intro sentence and I think it needs some other changes.  I suggest the following:

Certain scenarios require the use of a `Filter` implementation which can be applied to all applications deployed to Liberty.  For example, a `Filter` could be used to add new headers to each request. The `ServletContainerInitializer` interface provides a method called `onStartup` which can be used to add a `Filter` implementation class to all applications.

The BELLs feature can be used to configure a library that provides a `ServletContainerInitializer` implementation as a service.  The `ServletContainerInitializer` service `onStartup` method is called by Liberty for each Web Application that starts. In the following example, a `ServletContainerInitializer` implementation is configured  to add a `Filter` implementation with each `ServletContext` object. The `ServletContext.addFilter(String, Filter)` method is used by the `onStartup` method to add the `Filter` implementation:

**Comment by Rwalls1** [comment]
@lauracowen Please review: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/feature/bells-1.0.html

**Comment by lauracowen** [comment]
thanks both. Signing off

**Comment by Rwalls1** [comment]
@dmuelle Please peer review: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/feature/bells-1.0.html

**Comment by dmuelle** [comment]
## Peer review

examples look good- made a few suggestions for clarity and to remove passive voice

### Intro
- The metatype description says "Basic Extensions using Liberty Libraries (BELL)". I think throughout this doc, it should be "the BELL feature", or the metatype should be edited to say "Basic Extensions using Liberty Libraries (BELLs)". Either way, should be consistent.
- The first sentence is hard to read- I think because it says "...enables libraries to extend the Open Liberty runtime by allowing libraries to define.." Based on @tjwatson comment above,  I think you could simplify the sentence to:

"The BELL feature enables libraries to define the implementation of an Open Liberty SPI or API by using Java `ServiceLoader` configuration files."

### Configuring a Servlet Filter for applications
- I dont think servlet or or filter need to be capitalized.
- The first example is application code, the second is server config. The descriptions for each should mention where the code is- at least to make clear that it's not all in the same file

- throughout this section, `Filter`--> servlet filter. No need to monospace or cap filter, but specify "servlet filter" for clarity

- "Certain scenarios require the use of a `Filter` implementation that can be applied to all applications that are deployed to Open Liberty."---> 
 "In certain scenarios, a servlet filter implementation must be applied to all applications that are deployed to Open Liberty." 

- "The `ServletContainerInitializer` interface provides a method called `onStartup` that can be used to add a Filter implementation class to all applications." --> 
 "The `ServletContainerInitializer` interface provides an `onStartup` method,  which can add a servlet filter implementation class to all applications."

- "The `ServletContainerInitializer` service `onStartup` method that is called by Open Liberty for each application that starts."-->
- "Open Liberty calls the `onStartup` method of the ServletContainerInitializer` implementation  for each application that starts."

- "The `ServletContext.addFilter(String, Filter)` method is used by the `onStartup` method to add the Filter implementation:"
--->
"The `onStartup` method uses the `ServletContext.addFilter(String, Filter)` method to add the `myFilter` servlet filter implementation:


- I think this description needs work:
"To register the ServletContainerInitializer interface that adds the Servlet Filter for an application, a class name is specified in the library JAR file as shown in the following example:"

While the description is accurate for how the config works, it doesn't really describe the code example, which shows how to use Bells to configure the library reference.  I think the info you have is useful, but it also needs to describe what is shown in the example. We don't need to show the Java code from the Library JAR, but do need to explain the server xml config shown here. As is, it sounds like the code shows whats in the library, when really it just shows how the server references the library. 

### Adding the shared library for a custom user registry
- i think the title might be more clear if it was "Adding a custom user registry from a shared library", since the user registry is the value
- The following example adds the resources/sharedLib shared library that contains the bellsCur.jar JAR file that the BELLs feature uses to load the custom user registry: -->
The following example adds the resources/sharedLib shared library that contains the bellsCur.jar JAR file. The `bell` element references this library to load the custom user registry:

**Comment by Rwalls1** [comment]
@dmuelle Thanks, I've made those updates: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/feature/bells-1.0.html

**Comment by Rwalls1** [comment]
New link: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.2/reference/feature/bells-1.0.html

**Comment by Rwalls1** [comment]
Staging: https://staging-openlibertyio.mybluemix.net/docs/21.0.0.6/reference/feature/bells-1.0.html

**Comment by chirp1** [comment]
Hi Richard, Yet another nice set of config examples! Here are my comments:
- For "...to add a servlet filter implementation with each ServletContext object in a Java file.", I think you mean "....to add a servlet filter implementation for each ServletContext object that is in a Java file.". Change as appropriate. If my suggestion isn't right, would you let me know why not? Thanks!
- I believe that we switched from gerunds to commands for the section titles (ie Configuring > Configure; Adding > Add).  Check with David and change as appropriate. 
- Change "...path/to/jar.." to the appropriate way to highlight variables that we agreed on. Check with David as necessary.
- For "The following example adds the resources/sharedLib shared library that contains the bellsCur.jar JAR file in the server.xml file.", I'm confused by "in the server.xml file". Do you maybe mean that the example config snippet statements are in the server.xml file? Reword to make clear.

**Comment by Rwalls1** [comment]
@chirp1 Thanks, I've addressed your feedback: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.8/reference/feature/bells-1.0.html

**Comment by chirp1** [comment]
Hi Richard,

- I see that you changed "...path/to/jar.." by italicizing each word, but change "..._path/to/jar_.." to "..._path to jar_..". (Replace the forward slashes with spaces.)

**Comment by Rwalls1** [comment]
@chirp1 Thanks, I've addressed your latest feedback: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.8/reference/feature/bells-1.0.html

**Comment by chirp1** [comment]
Almost there! Change " to add the path/to/jar JAR file to the library:" to "...to add the path to jar JAR file to the library:".

**Comment by Rwalls1** [comment]
@chirp1 Thanks, please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.8/reference/feature/bells-1.0.html

**Comment by chirp1** [comment]
Update looks good! Moving to Ready to Publish.

**Comment by Rwalls1** [comment]
Staging: https://staging-openlibertyio.mybluemix.net/docs/21.0.0.8/reference/feature/bells-1.0.html

**Comment by Rwalls1** [comment]
Updates have been merged to Vnext and will publish with 21.0.0.8
